2020-01-Winter/3_stacks_queues/README.md

```diff
@@ -138,21 +168,54 @@ Go to [Top](#top)
 <!-- Don't remove -->
 <a name="s3"/>
 
-### 3. SOLUTION 3 TODO :bug:
-
-Source: TODO :bug:
-
-#### Naive/Simple Solution 
-
-TODO :bug:
-
-#### Optimal Solution
-
-TODO :bug:
-
-#### Testing The Solutions OR Driver For Solution
-
-TODO :bug:
+### 3. Trapping Rain Water 
+
+Source: [LeetCode](https://leetcode.com/problems/trapping-rain-water/)
+
+#### Solution
+
+We check the left most bar and right most bar, if the left most bar is smaller or equal to the right most bar, 
+we know that the water level on the right side of the left most bar should be at the left most bar's level. 
+At this point, we can check the height of our new left most bar without the right most bar. 
+We keep doing this until those 2 pointers meet each other.
+
+```python
+def trap(self, height):
+        s = 0
+        l, r = 0, len(height)-1
+        while l < r:
+            i = 1
+            if height[l] < height[r]:
+                while height[l] > height[l+i]:
+                    s += height[l] - height[l+i]
+                    i += 1
+                l += i
+            else:
+                while height[r] > height[r-i]:
+                    s += height[r] - height[r-i]
+                    i += 1
+                r -= i
+        return s
+ ```
```

Comment:
The writeup is good and we can leave it in, but what's missing is the solution which uses stacks. There are multiple ways of solving this problem, and with this problem, I'd like to illustrate how stacks can be used to keep track of the "wells" in the shape the data produces. (Also, the stack approach fits the topic of the week.)

I can add this later.